source/common/compressor/zlib_compressor_impl.cc

```diff
@@ -31,32 +31,38 @@ void ZlibCompressorImpl::init(CompressionLevel comp_level, CompressionStrategy c
   initialized_ = true;
 }
 
-void ZlibCompressorImpl::flush(Buffer::Instance& output_buffer) {
-  process(output_buffer, Z_SYNC_FLUSH);
-}
-
 uint64_t ZlibCompressorImpl::checksum() { return zstream_ptr_->adler; }
 
-void ZlibCompressorImpl::compress(const Buffer::Instance& input_buffer,
-                                  Buffer::Instance& output_buffer) {
-  const uint64_t num_slices = input_buffer.getRawSlices(nullptr, 0);
+void ZlibCompressorImpl::compress(Buffer::Instance& buffer, bool trailer) {
+  const uint64_t num_slices = buffer.getRawSlices(nullptr, 0);
   Buffer::RawSlice slices[num_slices];
-  input_buffer.getRawSlices(slices, num_slices);
+  buffer.getRawSlices(slices, num_slices);
 
   for (const Buffer::RawSlice& input_slice : slices) {
     zstream_ptr_->avail_in = input_slice.len_;
     zstream_ptr_->next_in = static_cast<Bytef*>(input_slice.mem_);
-    process(output_buffer, Z_NO_FLUSH);
+    process(buffer, Z_NO_FLUSH);
   }
+
+  buffer.drain(buffer.length());
```

Comment:
How can this be correct? Aren't you compressing into the buffer directly so maybe you should only drain the initial uncompressed data? I'm confused.

Reply:
@mattklein123 What happens is that all the slices are taken inside `zlib` without producing any output. Only when Z_SYNC_FLUSH or Z_FINISH is called, `zlib` updates `avail_out` and `next_out`. The problem would be if we call  
If the intention is to produce output on every slice, then `Z_SYNC_FLUSH` should be used in replacement of `Z_NO_FLUSH`. I'm not sure about the benefit of doing this though.

@dio One think that we could potentially simplify, are the number of helper functions.

Reply:
I see. OK can you add some comments? It's a bit confusing now.

Reply:
@gsagula somehow, following @mattklein123 suggestion fixed the failed test of the decompressor when using the compressor with uncommon params.